src/umockdev-ioctl.vala

```diff
@@ -535,13 +617,91 @@ public class IoctlClient : GLib.Object {
     }
 }
 
+/**
+ * UMockdevIoctlBaseClass:
+ * @handle_ioctl: Override ioctl emulation
+ * @handle_read: Override read emulation
+ * @handle_write: Override write_emulation
+ * @client_connected: A device was opened
+ * @client_vanished: A device was closed
+ *
+ * The base class for an device ioctl and read/write handling. You can either
+ * override the corresponding vfuncs or connect to the signals to customize
+ * the emulation.
+ *
+ * Since: 0.16
+ */
 
 /**
  * UMockdevIoctlBase:
  *
  * The #UMockdevIoctlBase class is a base class to emulate and record ioctl
  * operations of a client. It can be attached to an emulated device in the
  * testbed and will then be used.
+ *
+ * Since: 0.16
+ */
+/**
+ * UMockdevIoctlBase::handle-ioctl:
+ * @handler: A #UMockdevIoctlBase
+ * @client: A #UMockdevIoctlClient
+ *
+ * Called when an ioctl is requested by the client.
+ *
+ * Access the #UMockdevIoctlClient::arg property of @client to retrieve the
+ * argument of the ioctl. This is a pointer sized buffer initially with the
+ * original argument passed to the ioctl. If this is pointing to a struct, use
+ * umockdev_ioctl_data_resolve() to retrieve the underlying memory and update
+ * the pointer. Resolve any futher pointers in the structure in the same way.
+ *
+ * After resolving the memory, you can access it as if it was local. Do not
+ * forget to call umockdev_ioctl_data_dirty() if you modify the memory.
```

Comment:
Would there be much harm in always syncing back the memory in `client_complete()`, instead of tracking it with dirty explicitly? It's not like this ioctl emulation would be anywhere near "fast" :grin: 

I.e. calling `resolve` could set an internal flag that the `arg` is a pointer, and `complete` would then do the syncing.

Reply:
We could do, it generates a few more context switches in cases where it isn't needed. Though, if we want to avoid that we could even create a backing buffer for the original client side data and check if anything has actually changed.

So, fine with me to remove this optimization and simplify the API that way.

Reply:
API-wise it would be more robust, but admittedly I can't judge how bad it would affect performance, and if that matters at all. Mostly a question, I'll let you decide.

Reply:
There are two different ways we can optimize this. i.e. `memcpy` and changing the protocol to avoid some of the round-trips. So, I think the right thing here is to make it implicit and remove it from the API.

This will also mean loading the read buffer from the application. But, not really a big concern.

At the end, I am mostly concerned about having unnecessary context switches. These seem to cost us a lot of time if there is CPU contention on the system (I suppose the scheduler is not fair with regard to overall CPU time in this case).

Reply:
OK, did the changes. That does simplify the ioctl handling code quite a bit throughout to be honest.

The performance impact should be pretty small overall.